### PR TITLE
implemented exceptions in Controller, Menu and Customer's Menu

### PR DESCRIPTION
**Pull request by dk25**
Implemented exceptions in:
- Menu (general checks on validity of menu options)
- Controller (checks for nextInt and nextString)
- Buy ticket scenario;

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/Espress0JavaTeam/AirLine/pull/9" target="_blank">ReviewNinja

File: src/com/acme/command/impl/flight/BuyTicket.java [hunk 1]
```diff
@@ -24,19 +24,38 @@ public void execute() {
...
+
+            } catch (RuntimeException e) {
+                ErrorMessage.notValidFlightId();
+            }
+        }
+        return id;
+    }
+
+    private boolean isValidFlightId(int id) {
+        return (id == 0) | fs.hasFlight(id);
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/command/impl/flight/BuyTicket.java]: do you know - it is binary OR not logical
https://www.screencast.com/t/wxIe6uFn

File: src/com/acme/command/impl/flight/BuyTicket.java [hunk 2]
```diff
@@ -24,19 +24,38 @@ public void execute() {
...
+        int id = -1;
+
         PromptMessage.flightId();
-        return Controller.nextInt();
+
+        while(!isValidFlightId(id)) {
+            try {
+                id = Controller.nextInt();
+
+                if (!isValidFlightId(id)) {
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/command/impl/flight/BuyTicket.java]: please avoid this double-check

File: src/com/acme/command/impl/flight/BuyTicket.java [hunk 3]
```diff
@@ -24,19 +24,38 @@ public void execute() {
...
             return;
         }
 
-        flight = fs.getFlightBy(flightID);
+        flight = fs.getFlightBy(flightId);
         flight.sellTicket();
     }
 
-    private int askFlightID() {
+    private int askFlightId() {
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/command/impl/flight/BuyTicket.java]: allow method to throw exception and catch/process it in execute()

File: src/com/acme/view/Menu.java [hunk 4]
```diff
@@ -40,25 +41,34 @@ private void showMenu() {
...
-        return Controller.nextInt();
+
+        int selection = -1;
+
+        while(!isValidCommand(selection)) {
+            PromptMessage.menu();
+            try {
+                selection = Controller.nextInt();
+
+                if (!isValidCommand(selection)) {
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/view/Menu.java]: again double-check

File: src/com/acme/view/Menu.java [hunk 5]
```diff
@@ -40,25 +41,34 @@ private void showMenu() {
...
-            cmd.execute();
-        } else {
-            ErrorMessage.wrongInput();
-        }
+        this.commands.get(selection).execute();
     }
 
-    private boolean isValid(int selection) {
-        return this.commands.containsKey(selection);
+    private boolean isValidCommand(int selection) {
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/view/Menu.java]: allow isValidCommand to throw exception and catch it in method which calls it

File: src/com/acme/controller/Controller.java [hunk 6]
```diff
@@ -31,7 +31,7 @@ public static String nextStr() {
                 str = scanner.nextLine();
 
                 if (str == null || str.isEmpty()) {
-                    throw new RuntimeException("String is empty");
+                    throw new RuntimeException("String is empty"); //  change illegal argument exception, illegal format
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/controller/Controller.java]: чому не змінив?

File: src/com/acme/controller/Controller.java [hunk 7]
```diff
@@ -24,21 +24,20 @@ public static int nextInt() {
...
+        }
 
-                return str;
+        return str;
+    }
 
-            } catch (RuntimeException e) {
-                ErrorMessage.notString();
-            }
+    private static void validateString(String str) throws InvalidStringException {
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/controller/Controller.java]: better StringIsEmptyException

File: src/com/acme/exception/InvalidStringException.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.acme.exception;
+
+import com.acme.localisation.Messages.ErrorMessage;
+
+public class InvalidStringException extends AirlineException {
+
+    public InvalidStringException() {
+        ErrorMessage.notString();
```
Comment (VolodymyrBondarchuk) [inline src/com/acme/exception/InvalidStringException.java]: ErrorMessage.stringIsEmpty()

**Pull request merged**